src/rpc/rawtransaction.cpp

```diff
@@ -893,6 +901,22 @@ UniValue sendrawtransaction(const JSONRPCRequest& request)
     {
         pnode->PushInventory(inv);
     });
+
+    } // cs_main
+
+    if (should_wait_on_validationinterface) {
+        // If wallet is enabled, ensure that the wallet has been made aware
+        // of the new transaction prior to returning. This prevents a race
+        // where a user might call sendrawtransaction with a transaction
+        // to/from their wallet, immediately call some wallet RPC, and get
+        // a stale result because callbacks have not yet been processed.
+        std::promise<void> promise;
+        CallFunctionInValidationInterfaceQueue([&promise] {
+            promise.set_value();
+        });
+        promise.get_future().wait();
```

Comment:
In commit "Fix wallet RPC race by waiting for callbacks in sendrawtransaction"

This seems right. Possible tweaks:

* `should_wait_on_validationinterface` seems like it is always true at this point, maybe the variable is not needed.
* Might be better to call `CallFunctionInValidationInterfaceQueue` before releasing cs_main, to avoid waiting for notifications that might be queued in the meantime that we don't care about.
* Might be cheaper / simpler just to hold on to the tx reference (remove std::move) instead of copying the hash to a temporary variable.
* Maybe release cs_main before the PushInventory loop.